### DISCUSSION TITLE
use custom filter to dynamic show/hide column, the performance is too low!!

### DISCUSSION THREAD

**Discussion by CoderAWei** [ant-design/ant-design#44244] [General]
### Reproduction link

[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/zi-ding-yi-shai-xuan-de-sou-suo-antd-4-24-13-forked-d6zrtl?file=/demo.tsx)

### Steps to reproduce

- Click filter icon and hover ``COLOMNS``
- Select one/multiple columns to show/hide, and you will find that the corresponding speed of the page is very slow or even stuck. 

### What is expected?

I hope the page can respond quickly

### What is actually happening?

respond very slow!

| Environment | Info |
| --- | --- |
| antd | 4.24.8 |
| React | 18.2.0 |
| System | Windows |
| Browser | Google Chrome115.0.5790.102 |

---
使用的方法是通过css的display属性来显示隐藏，也尝试过使用splice改变column源数据，也是一样的很卡。不知道造成这样的原因是什么，可以帮忙看看吗或者针对这种需求还有没有更好的解决方案，谢谢！

**Comment by MadCcc** [comment]
槽点有点多，挑几个刺：
1. 不要在组件里定义组件
2. 代码里变量的作用域相当混乱

给个修复后的 demo，看看是不是你的预期
https://codesandbox.io/s/zi-ding-yi-shai-xuan-de-sou-suo-antd-4-24-13-forked-847flr?file=/demo.tsx
